README.md



基于搜狗微信搜索的微信公众号爬虫接口
===
#####Tips:目前该接口仅仅抓取公众号信息和公众号站内搜索
```
 __        __        _           _   ____
 \ \      / /__  ___| |__   __ _| |_/ ___|  ___   __ _  ___  _   _
  \ \ /\ / / _ \/ __| '_ \ / _` | __\___ \ / _ \ / _` |/ _ \| | | |
   \ V  V /  __/ (__| | | | (_| | |_ ___) | (_) | (_| | (_) | |_| |
    \_/\_/ \___|\___|_| |_|\__,_|\__|____/ \___/ \__, |\___/ \__,_|
                                                 |___/
```

# 项目简介
基于搜狗微信搜索的微信公众号爬虫接口，可以扩展成基于搜狗搜索的爬虫
目前仅支持公众号信息查询和公众号内关键之文章搜索

如果有问题，请提issue
---
# TODO
- [x] 公众号信息
- [ ] 主页热门公众号获取
- [ ] 文章详情页信息
- [x] 账号内搜索
- [ ] 验证码识别
- [ ] 兼容py2



news.txt
安永EY,安永EY,EY_GreaterChina
安永EY,安永EY,EY_GreaterChina
毕马威KPMG,毕马威KPMG,KPMGChina
文思海辉,文思海辉,pactera2012
TestinYunce,Testin云测,TestinChina
pactera2012,文思海辉,pactera2012
ThoughtWorks,ThoughtWorks,TW-Insights
omcube,橙色魔方,omcube
聚合科技集团,福建省云,ssddzd
车300二手车,车300二手车,chesanbai
数美反欺诈,数美反欺诈,ishumei2015
停车百事通,停车百事通,gh_5f6cb10f4605
analysys易观,Analysys易观,enfodesk
油客网,油客网,gh_1491d5dc54bb
TemenosChina,金融TE客,TemenosChina
财富共赢俱乐部,未来,wlcf-egd
握奇,握奇,Watchdata_CN
文思海辉金融,文思海辉金融,gh_38b445cac409
金蝶,金蝶,kingdee_268
西安葡萄城,西安葡萄城,GrapeCityXA
毕马威KPMG,毕马威KPMG,KPMGChina
红帽,红帽,redhatchina
ThoughtWorks洞见,ThoughtWorks洞见,TW-Insights
FIT2CLOUD,FIT2CLOUD,FIT2CLOUD-CN
AWS,AWS,awschina
(EasyOps)",EasyOps,gh_a8518ebee166
思特沃克,思特沃克,thoughtworks-school
haizhixingtu,海致星图er,haizhixingtu
博思云为,博思云为,bosiyunwei
IBM中国,IBM中国,IBMGCG
中指研究院,中指研究院,indexsoufun
云估价,云估价,yungujia
国信达数据,国信达数据,gxddata
网上管家婆,网上管家婆,wsgjp_
"公众号：,湖心亭看雪客,huxintingkanxueke
文思海辉金融",文思海辉金融,gh_38b445cac409
"公众号：,湖心亭看雪客,huxintingkanxueke
诺诺金服,航信,gh_436e7902a6d0
时代银通,时代银通,ERA_YINTONG
zysjguanfang,中邮世纪,zysjguanfang
BANGCLE,梆梆安全,BANGCLE
"神州泰岳,神州泰岳,gh_ea2cbb12a589
追一科技,追一科技,zhuiyikeji
视野金服,视野金服,seeyiicom
梆梆安全,梆梆安全,BANGCLE
tianrang-sh,天壤,tianrang-sh
中指研究院,中指研究院,indexsoufun
云估价,云估价,yungujia
国信达数据,国信达数据,gxddata
网上管家婆,网上管家婆,wsgjp_
"公众号：,湖心亭看雪客,huxintingkanxueke
文思海辉金融",文思海辉金融,gh_38b445cac409
"公众号：,湖心亭看雪客,huxintingkanxueke
诺诺金服,航信,gh_436e7902a6d0
时代银通,时代银通,ERA_YINTONG
芒果TV,芒果TV,jinyingwang
易道博识,易道博识,yidaoboshi
吉大正元,吉大正元,JITweixin
江南科友,江南科友,keyou_cn
zysjguanfang,中邮世纪,zysjguanfang
BANGCLE,梆梆安全,BANGCLE
"神州泰岳,神州泰岳,gh_ea2cbb12a589
追一科技,追一科技,zhuiyikeji
视野金服,视野金服,seeyiicom
梆梆安全,梆梆安全,BANGCLE
tianrang-sh,天壤,tianrang-sh
腾讯云,腾讯云,TencentCloud
iFLYTEK1999,科大讯飞,iFLYTEK1999
nanjingguiyu,硅基智能机器人,nanjingguiyu
微软科技,微软科技,mstech2014
神州泰岳,神州泰岳,gh_ea2cbb12a589
云孚科技,云孚科技,gh_527b63396240
Teradata天睿大数据分析,Teradata天睿大数据分析,TeradataBigData
Kyligence,Kyligence,Kyligence
微策略商业智能,微策略,gh_9db1fc59672b
CognosCommunity,CognosCommunity,gh_5e7ca1788b5c
Tableau社区,Tableau社区,TableauChina
同方股份有限公司,同方股份有限公司,tongfangkeji
analysys易观,Analysys易观,enfodesk
"依登软件：,伊登软件,eden-soft
eden-soft",伊登软件,eden-soft


const.py
# -*- coding: utf-8 -*-

from functools import wraps

from exceptions import WechatSogouException

agents = [
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; AcooBrowser; .NET CLR 1.1.4322; .NET CLR 2.0.50727)",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.0; Acoo Browser; SLCC1; .NET CLR 2.0.50727; Media Center PC 5.0; .NET CLR 3.0.04506)",
    "Mozilla/4.0 (compatible; MSIE 7.0; AOL 9.5; AOLBuild 4337.35; Windows NT 5.1; .NET CLR 1.1.4322; .NET CLR 2.0.50727)",
    "Mozilla/5.0 (Windows; U; MSIE 9.0; Windows NT 9.0; en-US)",
    "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Win64; x64; Trident/5.0; .NET CLR 3.5.30729; .NET CLR 3.0.30729; .NET CLR 2.0.50727; Media Center PC 6.0)",
    "Mozilla/5.0 (compatible; MSIE 8.0; Windows NT 6.0; Trident/4.0; WOW64; Trident/4.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; .NET CLR 1.0.3705; .NET CLR 1.1.4322)",
    "Mozilla/4.0 (compatible; MSIE 7.0b; Windows NT 5.2; .NET CLR 1.1.4322; .NET CLR 2.0.50727; InfoPath.2; .NET CLR 3.0.04506.30)",
    "Mozilla/5.0 (Windows; U; Windows NT 5.1; zh-CN) AppleWebKit/523.15 (KHTML, like Gecko, Safari/419.3) Arora/0.3 (Change: 287 c9dfb30)",
    "Mozilla/5.0 (X11; U; Linux; en-US) AppleWebKit/527+ (KHTML, like Gecko, Safari/419.3) Arora/0.6",
    "Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.8.1.2pre) Gecko/20070215 K-Ninja/2.1.1",
    "Mozilla/5.0 (Windows; U; Windows NT 5.1; zh-CN; rv:1.9) Gecko/20080705 Firefox/3.0 Kapiko/3.0",
    "Mozilla/5.0 (X11; Linux i686; U;) Gecko/20070322 Kazehakase/0.4.5",
    "Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.9.0.8) Gecko Fedora/1.9.0.8-1.fc10 Kazehakase/0.5.6",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/535.11 (KHTML, like Gecko) Chrome/17.0.963.56 Safari/535.11",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_7_3) AppleWebKit/535.20 (KHTML, like Gecko) Chrome/19.0.1036.7 Safari/535.20",
    "Opera/9.80 (Macintosh; Intel Mac OS X 10.6.8; U; fr) Presto/2.9.168 Version/11.52",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/536.11 (KHTML, like Gecko) Chrome/20.0.1132.11 TaoBrowser/2.0 Safari/536.11",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.71 Safari/537.1 LBBROWSER",
    "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E; LBBROWSER)",
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; QQDownload 732; .NET4.0C; .NET4.0E; LBBROWSER)",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/535.11 (KHTML, like Gecko) Chrome/17.0.963.84 Safari/535.11 LBBROWSER",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E)",
    "Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E; QQBrowser/7.0.3698.400)",
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; QQDownload 732; .NET4.0C; .NET4.0E)",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; Trident/4.0; SV1; QQDownload 732; .NET4.0C; .NET4.0E; 360SE)",
    "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; QQDownload 732; .NET4.0C; .NET4.0E)",
    "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; .NET4.0E)",
    "Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.89 Safari/537.1",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.1 (KHTML, like Gecko) Chrome/21.0.1180.89 Safari/537.1",
    "Mozilla/5.0 (iPad; U; CPU OS 4_2_1 like Mac OS X; zh-cn) AppleWebKit/533.17.9 (KHTML, like Gecko) Version/5.0.2 Mobile/8C148 Safari/6533.18.5",
    "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:2.0b13pre) Gecko/20110307 Firefox/4.0b13pre",
    "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:16.0) Gecko/20100101 Firefox/16.0",
    "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.11 (KHTML, like Gecko) Chrome/23.0.1271.64 Safari/537.11",
    "Mozilla/5.0 (X11; U; Linux x86_64; zh-CN; rv:1.9.2.10) Gecko/20100922 Ubuntu/10.10 (maverick) Firefox/3.6.10",
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36",
]


def Const(cls):
    @wraps(cls)
    def new_setattr(self, name, value):
        raise WechatSogouException('const : {} can not be changed'.format(name))

    cls.__setattr__ = new_setattr
    return cls


@Const
class _WechatSogouSearchArticleTypeConst(object):
    all = 'all'
    rich = 'rich'
    video = 'video'
    image = 'image'


@Const
class _WechatSogouSearchArticleTimeConst(object):
    """搜索条件 时间

    0 没有限制 / 1一天 / 2一周 / 3一月 / 4一年 / 5自定
    """
    anytime = 0
    day = 1
    week = 2
    month = 3
    year = 4
    specific = 5


@Const
class _WechatSogouHotIndexConst(object):
    hot = 'hot'  # 热门
    gaoxiao = 'gaoxiao'  # 搞笑
    health = 'health'  # 养生
    sifanghua = 'sifanghua'  # 私房话
    gossip = 'gossip'  # 八卦
    technology = 'technology'  # 科技
    finance = 'finance'  # 财经
    car = 'car'  # 汽车
    life = 'life'  # 生活
    fashion = 'fashion'  # 时尚
    mummy = 'mummy'  # 辣妈 / 育儿
    travel = 'travel'  # 旅行
    job = 'job'  # 职场
    food = 'food'  # 美食
    history = 'history'  # 历史
    study = 'study'  # 学霸 / 教育
    constellation = 'constellation'  # 星座
    sport = 'sport'  # 体育
    military = 'military'  # 军事
    game = 'game'  # 游戏
    pet = 'pet'  # 萌宠


@Const
class _Const(object):
    hot_index = _WechatSogouHotIndexConst()
    search_article_type = _WechatSogouSearchArticleTypeConst()
    search_article_time = _WechatSogouSearchArticleTimeConst()


WechatSogouConst = _Const()


request.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, unicode_literals, print_function

import datetime
from collections import OrderedDict

from const import WechatSogouConst

_search_type_gzh = 1  # 公众号
_search_type_article = 2  # 文章


class WechatSogouRequest:
    @staticmethod
    def gen_search_article_url(keyword, page=1, timesn=WechatSogouConst.search_article_time.anytime,
                               article_type=WechatSogouConst.search_article_type.all, ft=None, et=None):
        """拼接搜索 文章 URL

        Parameters
        ----------
        keyword : str or unicode
            搜索文字
        page : int, optional
            页数 the default is 1
        timesn : WechatSogouConst.search_article_time
            时间 anytime 没有限制 / day 一天 / week 一周 / month 一月 / year 一年 / specific 自定
            默认是 anytime
        article_type : WechatSogouConst.search_article_type
            含有内容的类型 image 有图 / video 有视频 / rich 有图和视频 / all 啥都有
            默认是 all
        ft, et : datetime.date
            当 tsn 是 specific 时，ft 代表开始时间，如： 2017-07-01
            当 tsn 是 specific 时，et 代表结束时间，如： 2017-07-15

        Returns
        -------
        str
            search_article_url
        """
        assert isinstance(page, int) and page > 0
        assert timesn in [WechatSogouConst.search_article_time.anytime,
                          WechatSogouConst.search_article_time.day,
                          WechatSogouConst.search_article_time.week,
                          WechatSogouConst.search_article_time.month,
                          WechatSogouConst.search_article_time.year,
                          WechatSogouConst.search_article_time.specific]

        if timesn == WechatSogouConst.search_article_time.specific:
            assert isinstance(ft, datetime.date)
            assert isinstance(et, datetime.date)
            assert ft <= et
        else:
            ft = ''
            et = ''

        interation_image = 458754
        interation_video = 458756
        if article_type == WechatSogouConst.search_article_type.rich:
            interation = '{},{}'.format(interation_image, interation_video)
        elif article_type == WechatSogouConst.search_article_type.image:
            interation = interation_image
        elif article_type == WechatSogouConst.search_article_type.video:
            interation = interation_video
        else:
            interation = ''

        qs_dict = OrderedDict()
        qs_dict['type'] = _search_type_article
        qs_dict['page'] = page
        qs_dict['ie'] = 'utf8'
        qs_dict['query'] = keyword
        qs_dict['interation'] = interation
        if timesn != 0:
            qs_dict['tsn'] = timesn
            qs_dict['ft'] = str(ft)
            qs_dict['et'] = str(et)

        # TODO 账号内搜索
        # '账号内 http://weixin.sogou.com/weixin?type=2&ie=utf8&query=%E9%AB%98%E8%80%83&tsn=3&ft=&et=&interation=458754
        # &wxid=oIWsFt1tmWoG6vO6BcsS7St61bRE&usip=nanhangqinggong'
        # qs['wxid'] = wxid
        # qs['usip'] = usip

        return 'http://weixin.sogou.com/weixin?{}'.format(urlencode(qs_dict))

    @staticmethod
    def gen_search_gzh_url(keyword, page=1):
        """拼接搜索 公众号 URL

        Parameters
        ----------
        keyword : str or unicode
            搜索文字
        page : int, optional
            页数 the default is 1

        Returns
        -------
        str
            search_gzh_url
        """
        assert isinstance(page, int) and page > 0

        qs_dict = OrderedDict()
        qs_dict['type'] = _search_type_gzh
        qs_dict['page'] = page
        qs_dict['ie'] = 'utf8'
        qs_dict['query'] = keyword

        return 'http://weixin.sogou.com/weixin?{}'.format(urlencode(qs_dict))

    @staticmethod
    def gen_hot_url(hot_index, page=1):
        """拼接 首页热门文章 URL

        Parameters
        ----------
        hot_index : WechatSogouConst.hot_index
            首页热门文章的分类（常量）：WechatSogouConst.hot_index.xxx
        page : int
            页数

        Returns
        -------
        str
            热门文章分类的url
        """

        assert hasattr(WechatSogouConst.hot_index, hot_index)
        assert isinstance(page, int) and page > 0

        index_urls = {
            WechatSogouConst.hot_index.hot: 0,  # 热门
            WechatSogouConst.hot_index.gaoxiao: 1,  # 搞笑
            WechatSogouConst.hot_index.health: 2,  # 养生
            WechatSogouConst.hot_index.sifanghua: 3,  # 私房话
            WechatSogouConst.hot_index.gossip: 4,  # 八卦
            WechatSogouConst.hot_index.technology: 5,  # 科技
            WechatSogouConst.hot_index.finance: 6,  # 财经
            WechatSogouConst.hot_index.car: 7,  # 汽车
            WechatSogouConst.hot_index.life: 8,  # 生活
            WechatSogouConst.hot_index.fashion: 9,  # 时尚
            WechatSogouConst.hot_index.mummy: 10,  # 辣妈 / 育儿
            WechatSogouConst.hot_index.travel: 11,  # 旅行
            WechatSogouConst.hot_index.job: 12,  # 职场
            WechatSogouConst.hot_index.food: 13,  # 美食
            WechatSogouConst.hot_index.history: 14,  # 历史
            WechatSogouConst.hot_index.study: 15,  # 学霸 / 教育
            WechatSogouConst.hot_index.constellation: 16,  # 星座
            WechatSogouConst.hot_index.sport: 17,  # 体育
            WechatSogouConst.hot_index.military: 18,  # 军事
            WechatSogouConst.hot_index.game: 19,  # 游戏
            WechatSogouConst.hot_index.pet: 20,  # 萌宠

        }
        return 'http://weixin.sogou.com/wapindex/wap/0612/wap_{}/{}.html'.format(index_urls[hot_index], page - 1)


identify_image.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, unicode_literals, print_function

# ws_cache = WechatCache()
import io
import time

import requests
from PIL import Image

# from chaojiying import Chaojiying_Client
# from wechatsogou.exceptions import WechatSogouVcodeOcrException
# from wechatsogou.filecache import WechatCache
# from wechatsogou.five import readimg, input
from exceptions import WechatSogouVcodeOcrException
from filecache import WechatCache

ws_cache = WechatCache()


def identify_image_callback_by_hand(im: bytes):
    """im：图片数据流"""
    im = Image.open(io.BytesIO(im))
    im.show()
    code = input("请输入验证码：")
    return code
def unlock_weixin_callback_example(url, req, resp, img, identify_image_callback):
    """手动打码解锁

    Parameters
    ----------
    url : str or unicode
        验证码页面 之前的 url
    req : requests.sessions.Session
        requests.Session() 供调用解锁
    resp : requests.models.Response
        requests 访问页面返回的，已经跳转了
    img : bytes
        验证码图片二进制数据
    identify_image_callback : callable
        处理验证码函数，输入验证码二进制数据，输出文字，参见 identify_image_callback_example

    Returns
    -------
    dict
        {
            'ret': '',
            'errmsg': '',
            'cookie_count': '',
        }
    """
    # no use resp

    unlock_url = 'https://mp.weixin.qq.com/mp/verifycode'
    data = {
        'cert': time.time() * 1000,
        'input': identify_image_callback(img)
    }
    headers = {
        'Host': 'mp.weixin.qq.com',
        'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        'Referer': url
    }
    r_unlock = req.post(unlock_url, data, headers=headers)
    if not r_unlock.ok:
        raise WechatSogouVcodeOcrException(
            'unlock[{}] failed: {}[{}]'.format(unlock_url, r_unlock.text, r_unlock.status_code))

    return r_unlock.json()

def identify_image_by_hand(im: bytes):
    """im：图片数据流"""
    im = Image.open(io.BytesIO(im))
    im.show()
    code = input("请输入验证码：")
    return code


def unlock_sogou_callback_example(url, req, resp, img, identify_image_callback):  # 微信公众号查询打码接口
    """手动打码解锁
    Parameters
    ----------
    url : str or unicode
        验证码页面 之前的 url
    req : requests.sessions.Session
        requests.Session() 供调用解锁
    resp : requests.models.Response
        requests 访问页面返回的，已经跳转了
    img : bytes
        验证码图片二进制数据
    identify_image_callback : callable
        处理验证码函数，输入验证码二进制数据，输出文字，参见 identify_image_callback_example

    Returns
    -------
    dict
        {
            'code': '',
            'msg': '',
        }
    """
    # no use resp
    url_quote = url.split('weixin.sogou.com/')[-1]
    unlock_url = 'http://weixin.sogou.com/antispider/thank.php'
    data = {
        'c': identify_image_by_hand(im=img),  # todo 打码验证 直接返回打码结果 img 验证码 流
        'r': url_quote,
        'v': 5
    }
    headers = {
        'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        'Referer': 'http://weixin.sogou.com/antispider/?from=' + url_quote
    }
    r_unlock = req.post(unlock_url, data, headers=headers)
    r_unlock.encoding = 'utf-8'
    if not r_unlock.ok:
        raise Exception(
            'unlock[{}] failed: {}'.format(unlock_url, r_unlock.text, r_unlock.status_code))
    return r_unlock.json()


def unlock_weixin_callback_example(url, req, resp, img, identify_image_callback):
    """手动打码解锁

    Parameters
    ----------
    url : str or unicode
        验证码页面 之前的 url
    req : requests.sessions.Session
        requests.Session() 供调用解锁
    resp : requests.models.Response
        requests 访问页面返回的，已经跳转了
    img : bytes
        验证码图片二进制数据
    identify_image_callback : callable
        处理验证码函数，输入验证码二进制数据，输出文字，参见 identify_image_callback_example

    Returns
    -------
    dict
        {
            'ret': '',
            'errmsg': '',
            'cookie_count': '',
        }
    """
    # no use resp

    unlock_url = 'https://mp.weixin.qq.com/mp/verifycode'
    data = {
        'cert': time.time() * 1000,
        'input': identify_image_callback(img)
    }
    headers = {
        'Host': 'mp.weixin.qq.com',
        'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        'Referer': url
    }
    r_unlock = req.post(unlock_url, data, headers=headers)
    if not r_unlock.ok:
        raise WechatSogouVcodeOcrException(
            'unlock[{}] failed: {}[{}]'.format(unlock_url, r_unlock.text, r_unlock.status_code))

    return r_unlock.json()


api.py
# coding=utf-8
import random
import re
import time
from urllib.parse import urlencode

import requests
from lxml import etree

from const import agents, _WechatSogouSearchArticleTimeConst
from exceptions import WechatSogouRequestsException
from exceptions import WechatSogouVcodeOcrException
from identify_image import identify_image_callback_by_hand
from identify_image import unlock_sogou_callback_example
from identify_image import ws_cache

index_url = "https://weixin.sogou.com/"


class WechatSogouAPI:
    def __init__(self, captcha_break_time=1, headers=None, **kwargs):
        assert isinstance(captcha_break_time, int) and 0 < captcha_break_time < 20

        self.captcha_break_times = captcha_break_time
        self.requests_kwargs = kwargs
        self.headers = headers
        if self.headers:
            self.headers['User-Agent'] = random.choice(agents)

        else:
            self.headers = {'User-Agent': random.choice(agents)}
        self.gzh_query_url = "https://weixin.sogou.com/weixin?{}"
        self.base_url = "https://weixin.sogou.com/weixin"
        self.index_url = "https://weixin.sogou.com/"

    def __set_cookie(self, suv=None, snuid=None, referer=None):
        suv = ws_cache.get('suv') if suv is None else suv
        snuid = ws_cache.get('snuid') if snuid is None else snuid
        _headers = {'Cookie': 'SUV={};SNUID={};'.format(suv, snuid)}
        if referer is not None:
            _headers['Referer'] = referer
        return _headers

    def __get(self, url, session, headers):
        h = {}
        if headers:
            for k, v in headers.items():
                h[k] = v
        if self.headers:
            for k, v in self.headers.items():
                h[k] = v
        if url:
            resp = session.get(url, headers=h, **self.requests_kwargs)
        else:
            raise Exception("__get:url为空", )
        if not resp.ok:
            raise WechatSogouRequestsException('WechatSogouAPI get error', resp)
        return resp

    def __set_cache(self, suv, snuid):
        ws_cache.set('suv', suv)
        ws_cache.set('snuid', snuid)

    def __unlock_sogou(self, url, resp, session, unlock_callback=None, identify_image_callback=None):
        if unlock_callback is None:
            unlock_callback = unlock_sogou_callback_example
        millis = int(round(time.time() * 1000))
        r_captcha = session.get('http://weixin.sogou.com/antispider/util/seccode.php?tc={}'.format(millis), headers={
            'Referer': url,
        })
        if not r_captcha.ok:
            raise WechatSogouRequestsException('WechatSogouAPI get img', r_captcha)
        r_unlock = unlock_callback(url, session, resp, r_captcha.content, identify_image_callback)
        if r_unlock['code'] != 0:
            raise WechatSogouVcodeOcrException(
                '[WechatSogouAPI identify image] code: {code}, msg: {msg}'.format(code=r_unlock.get('code'),
                                                                                  msg=r_unlock.get('msg')))
        else:
            self.__set_cache(session.cookies.get('SUID'), r_unlock['id'])

    def __get_by_unlock(self, url, referer=None, unlock_platform=None, unlock_callback=None,
                        identify_image_callback=None, session=None):
        assert unlock_platform is None or callable(unlock_platform)
        if identify_image_callback is None:
            identify_image_callback = identify_image_callback_by_hand
        assert unlock_callback is None or callable(unlock_callback)
        assert callable(identify_image_callback)

        if not session:
            session = requests.session()
        resp = self.__get(url, session, headers=self.__set_cookie(referer=referer))
        resp.encoding = 'utf-8'
        if 'antispider' in resp.url or '请输入验证码' in resp.text:
            for i in range(self.captcha_break_times):
                try:
                    unlock_platform(url=url, resp=resp, session=session, unlock_callback=unlock_callback,
                                    identify_image_callback=identify_image_callback)
                    break
                except WechatSogouVcodeOcrException as e:
                    if i == self.captcha_break_times - 1:
                        raise WechatSogouVcodeOcrException(e)

            if '请输入验证码' in resp.text:
                headers = self.__set_cookie(referer=referer)
                headers['User-Agent'] = 'Mozilla/5.0 (Windows NT 6.1; WOW64)'
                resp = self.__get(url, session, headers)
                resp.encoding = 'utf-8'
        return resp

    @staticmethod
    def _paras_gzh_list(resp):
        gzh_list = []
        html = etree.HTML(resp.text)
        query_list = html.xpath("//ul[@class='news-list2']/li")
        for item in query_list:
            data = {}
            try:
                data["wx_pub_id"] = "".join(
                    item.xpath(".//p[@class='info']/label[@name='em_weixinhao']/text()"))  # 微信号
                data["wx_name"] = "".join(item.xpath(".//p[@class='tit']/a//text()"))  # 公众号名称
                data["wx_query_id"] = "".join(item.xpath(".//@d"))  # 查询ID
                data["wx_identify_company"] = "".join(
                    [i.strip() for i in item.xpath(".//i[@class='identify']/parent::dd/text()")])  # 认证公司名称
                data["recently_article"] = "".join(item.xpath(".//a[@uigs]/text()"))  # 最近文章title
                # data["source_url"] = "" # todo 解析文章链接
                gzh_list.append(data)
            except Exception as e:
                raise KeyError("页面错误", resp.content.decode("utf-8"), e.args)
        return gzh_list

    def get_wx_info(self, gzh_name: str, unlock_sogou_callback=None) -> list:
        """
        用于获取微信公众号相关信息,如果公众号名称或者ID确定,那么列表第一个就是要搜索的公众号

        :return:
        [
            {
            "wx_pub_id": 微信号
            "wx_name": 微信名称
            "wx_query_id": 私有ID
            "wx_identify_company": 微信认证公司
            "recently_article" 最近文章名称
            },
            {..}..
        ]
        """
        params = {"type": 1,
                  "query": gzh_name,
                  "ie": "utf-8",
                  "s_from": "input",
                  "_sug_": "n",
                  "_sug_type_": ""
                  }
        url = self.gzh_query_url.format(urlencode(params))
        if unlock_sogou_callback is None:
            unlock_sogou_callback = self.__unlock_sogou
        while url:
            resp = self.__get_by_unlock(url, referer=url, unlock_platform=unlock_sogou_callback)
            gzh_list = self._paras_gzh_list(resp)
            url = self._is_exist_next_page(resp=resp)
            yield gzh_list

    @staticmethod
    def paras_list(resp) -> list:
        articles = []
        resp.encoding = "utf-8"
        html = etree.HTML(resp.text)
        for item in html.xpath("//ul[@class='news-list']/li"):
            data = {}
            data["title"] = "".join([i.strip() for i in item.xpath(".//h3/a/text()")])
            data["brief"] = "".join([i.strip() for i in item.xpath(".//div[@class='txt-box']/p//text()")])
            data["pub_gzh"] = "".join([i.strip() for i in item.xpath(".//div[@class='s-p']/a//text()")])
            data["pub_time"] = time.strftime('%Y-%m-%d', time.localtime(int(item.xpath(".//div[@class='s-p']/@t")[0])))
            data["source_url"] = "".join(item.xpath(".//div[@class='txt-box']/h3/a/@data-share"))
            articles.append(data)
        return articles

    def query_article(self, keyword: str, gzh: str = None,
                      sec_id: str = None,
                      tsn: str = _WechatSogouSearchArticleTimeConst.anytime,
                      ft: str = None,
                      et: str = None,
                      interation: str = None,
                      unlock_sogou_callback=None) -> list:
        """
        通过私有ID或者公众号名称查询与该公众号有关的关键字相关文章

        :param keyword: 查询关键字
        :param gzh: 公众号名称
        :param sec_id: 微信私有ID
        :param tsn: 查询时间类型 1:一天内,2:一周内,3:一月内,4:一年内,5:自定义时间
        :param ft:自定义开始时间 format-eg:2019-06-08
        :param et:自定义结束时间
        :return:文章列表
        """
        url = "https://weixin.sogou.com/weixin?{}"
        if not sec_id and gzh:
            gzh_list = self.get_wx_info(gzh)
            sec_id = next(gzh_list)[0].get("wx_query_id") if gzh_list else None
        assert sec_id is not None
        if tsn is "5" and not ft and not et:
            raise KeyError("参数错误")
        elif tsn is "5" and ft and et:
            if not re.match("\d{4}-[01]\d-[01]\d", et) and not re.match("\d{4}-[01]\d-[01]\d", ft):
                raise KeyError("date error", ft, et)
        params = {
            "type": 2,
            "ie": "utf-8",
            "query": keyword,
            "wxid": sec_id,
            "usip": gzh,
            "tsn": tsn,
            "ft": ft,
            'et': et,
            "interation": interation
        }
        url = url.format(urlencode(params))
        if unlock_sogou_callback is None:
            unlock_sogou_callback = self.__unlock_sogou
        while url:
            resp = self.__get_by_unlock(url, referer=url, unlock_platform=unlock_sogou_callback)
            gzh_list = self.paras_list(resp)
            url = self._is_exist_next_page(resp=resp)
            yield gzh_list

    @staticmethod
    def get_article_content(url, session=None):
        session = requests.Session() if not session else session
        resp = session.get(url)
        return resp.text

    @staticmethod
    def _is_exist_next_page(resp):
        base_url = "https://weixin.sogou.com/weixin"
        if "下一页" in resp.text:
            return base_url + str(etree.HTML(resp.text).xpath("//a[@id='sogou_next']/@href")[0])
        else:
            return False


if __name__ == '__main__':
    t = WechatSogouAPI()
    KEY = ""
    while KEY != "quit":
        KEY = input("请输入查询关键字(quit退出):")
        gzh = input("请输入公众号:")
        temp = t.query_article(keyword=KEY, gzh=gzh)
        for item in temp:
            print(item)

"""
验证码之后,会重复一个页面的数据
"""
